# <div style='padding:25px;background-color:maroon;color:white;border-radius:4px;font-size:100%;text-align: center'> Insurance Analytics and Prediction<br></div>

# <div style='padding:5px;background-color:maroon;color:white;border-radius:2px;font-size:100%;text-align: center'>Data Cleaning<br></div>

## <span style="color:Aqua;"> Objective of the Project:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; The objective of this project is to leverage advanced analytics techniques, including classification,
regression, and clustering, to extract valuable insights from insurance data. By analyzing a
comprehensive dataset, the project aims to enhance decision-making processes, optimize risk
assessment, and improve overall operational efficiency within the insurance industry.

### <p style="color:Aqua;"> Key Components:</p>

<p style="color:Tomato;font-size: 110%"> <b> 1. Customer Segmentation (Clustering):</b> </p>

<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Apply clustering algorithms to group policyholders based on similar characteristics and behavior.

<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Identify customer segments with common insurance needs and preferences.

<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Tailor marketing strategies and product offerings to specific clusters, enhancing customer engagement and increasing cross-selling opportunities.

<p style="color:Tomato;font-size: 110%"> <b> 2. Fraudulent or Legitimate Assessment (Classification):</b> </p>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span>  Implement a classification model to categorize insurance claims into predefined classes, such as
fraudulent or legitimate. <br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Utilize machine learning algorithms to predict the likelihood of a claim being fraudulent based on
historical data. <br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Enhance fraud detection capabilities to reduce financial losses and improve the accuracy of claim
assessments.

<p style="color:Tomato;font-size: 110%"> <b> 3. Premium Prediction (Regression):</b> </p>

<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Develop regression models to predict insurance premium pricing based on various factors such as age, location, coverage type, and previous claims history. <br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Explore the relationship between different variables and premiums to optimize pricing strategies. <br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Provide recommendations for personalized premium adjustments, leading to improved customer
satisfaction and retention.

## <span style="color:Aqua;">Importing libraries from Python</span>

In [85]:
import pandas as pd 
import os
import numpy as np
from IPython.display import display, HTML

pd.options.display.max_columns = 50
pd.set_option("display.precision", 4)
pd.set_option('display.float_format', '{:.4f}'.format)


## <span style="color:Aqua;">Data Exploration:

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> In this section we are gonna explore features of each variables (data columns) and understand any issues with the data which may affect our Machine Learning model (Predicting app).



### <span style="color:Tomato;">Reading Dataset from CSV:

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Reading data from the excel to pandas (Python's Data Wrangler).

In [86]:
df = pd.read_excel('insurance_data.xlsx')

### <span style="color:Tomato;"> Understanding the dataset:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the output below we see that we have 1000 rows and 39 columns. Next output is 5 sample data points generated randomly from the dataset.

In [87]:
display(HTML(f"<p style='color: orange; font-weight: bold;'>{df.shape}\n\n</p>"))
df.sample(5)

months_as_customer  age  policy_number policy_bind_date policy_state  \
807                 186   38         250833       2008-07-28           IN   
275                 243   43         367595       2006-02-03           IN   
876                 247   39         599031       1991-10-29           IN   
614                 131   33         432740       1990-10-09           IL   
411                  84   29         630998       2003-04-09           OH   

    policy_csl  policy_deductable  policy_annual_premium  umbrella_limit  \
807    250/500               2000              1347.3100               0   
275   500/1000                500              1307.7400               0   
876    100/300                500              1558.2900               0   
614    100/300               2000              1081.1700               0   
411    250/500               1000              1117.1700               0   

     insured_zip insured_sex insured_education_level insured_occupation  \
807       436467      FEMALE                      JD    protective-serv   
275       466137      FEMALE               Associate  machine-op-inspct   
876       456781      FEMALE                 Masters    protective-serv   
614       445120        MALE                      MD              sales   
411       473645      FEMALE             High School  machine-op-inspct   

    insured_hobbies insured_relationship  capital-gains  capital-loss  \
807         dancing            unmarried          80900       -111100   
275     board-games            own-child              0        -75700   
876         reading            unmarried              0        -49300   
614        yachting                 wife              0        -65200   
411     video-games        not-in-family              0        -29900   

    incident_date            incident_type   collision_type incident_severity  \
807    2015-02-02  Multi-vehicle Collision   Rear Collision      Minor Damage   
275    2015-01-28  Multi-vehicle Collision  Front Collision      Major Damage   
876    2015-02-16            Vehicle Theft                ?    Trivial Damage   
614    2015-01-28               Parked Car                ?      Minor Damage   
411    2015-02-12               Parked Car                ?    Trivial Damage   

    authorities_contacted incident_state incident_city   incident_location  \
807             Ambulance             NY   Springfield     1371 Texas Lane   
275             Ambulance             SC     Riverwood        6451 1st Hwy   
876                Police             WV   Springfield    2577 Texas Ridge   
614                Police             NY     Northbend  3246 Britain Ridge   
411                Police             SC     Arlington       3693 Pine Ave   

     incident_hour_of_the_day  number_of_vehicles_involved property_damage  \
807                         1                            3              NO   
275                        10                            3               ?   
876                         5                            1             YES   
614                         3                            1               ?   
411                         6                            1             YES   

     bodily_injuries  witnesses police_report_available  total_claim_amount  \
807                2          3                       ?               73800   
275                0          1                      NO               37530   
876                1          2                       ?                7700   
614                0          1                      NO                4900   
411                2          0                     YES                6820   

     injury_claim  property_claim  vehicle_claim auto_make auto_model  \
807         12300           12300          49200      Audi         A3   
275          4170            4170          29190      Jeep   Wrangler   
876           770            1540           5390      Saab         93   
614

In [88]:
df.head()

months_as_customer  age  policy_number policy_bind_date policy_state  \
0                 328   48         521585       2014-10-17           OH   
1                 228   42         342868       2006-06-27           IN   
2                 134   29         687698       2000-09-06           OH   
3                 256   41         227811       1990-05-25           IL   
4                 228   44         367455       2014-06-06           IL   

  policy_csl  policy_deductable  policy_annual_premium  umbrella_limit  \
0    250/500               1000              1406.9100               0   
1    250/500               2000              1197.2200         5000000   
2    100/300               2000              1413.1400         5000000   
3    250/500               2000              1415.7400         6000000   
4   500/1000               1000              1583.9100         6000000   

   insured_zip insured_sex insured_education_level insured_occupation  \
0       466132        MALE                      MD       craft-repair   
1       468176        MALE                      MD  machine-op-inspct   
2       430632      FEMALE                     PhD              sales   
3       608117      FEMALE                     PhD       armed-forces   
4       610706        MALE               Associate              sales   

  insured_hobbies insured_relationship  capital-gains  capital-loss  \
0        sleeping              husband          53300             0   
1         reading       other-relative              0             0   
2     board-games            own-child          35100             0   
3     board-games            unmarried          48900        -62400   
4     board-games            unmarried          66000        -46000   

  incident_date             incident_type   collision_type incident_severity  \
0    2015-01-25  Single Vehicle Collision   Side Collision      Major Damage   
1    2015-01-21             Vehicle Theft                ?      Minor Damage   
2    2015-02-22   Multi-vehicle Collision   Rear Collision      Minor Damage   
3    2015-01-10  Single Vehicle Collision  Front Collision      Major Damage   
4    2015-02-17             Vehicle Theft                ?      Minor Damage   

  authorities_contacted incident_state incident_city  incident_location  \
0                Police             SC      Columbus     9935 4th Drive   
1                Police             VA     Riverwood       6608 MLK Hwy   
2                Police             NY      Columbus  7121 Francis Lane   
3                Police             OH     Arlington   6956 Maple Drive   
4                   NaN             NY     Arlington       3041 3rd Ave   

   incident_hour_of_the_day  number_of_vehicles_involved property_damage  \
0                         5                            1             YES   
1                         8                            1               ?   
2                         7                            3              NO   
3                         5                            1               ?   
4                        20                            1              NO   

   bodily_injuries  witnesses police_report_available  total_claim_amount  \
0                1          2                     YES               71610   
1                0          0                       ?                5070   
2                2          3                      NO               34650   
3                1          2                      NO               63400   
4                0          1                      NO                6500   

   injury_claim  property_claim  vehicle_claim  auto_make auto_model  \
0          6510           13020          52080       Saab        92x   
1           780             780           3510   Mercedes       E400   
2          7700            3850          23100      Dodge        RAM   
3          6340            6340          50720  Chevrolet      Tahoe   
4          1300             6

In [89]:
df.tail()

months_as_customer  age  policy_number policy_bind_date policy_state  \
995                   3   38         941851       1991-07-16           OH   
996                 285   41         186934       2014-01-05           IL   
997                 130   34         918516       2003-02-17           OH   
998                 458   62         533940       2011-11-18           IL   
999                 456   60         556080       1996-11-11           OH   

    policy_csl  policy_deductable  policy_annual_premium  umbrella_limit  \
995   500/1000               1000              1310.8000               0   
996    100/300               1000              1436.7900               0   
997    250/500                500              1383.4900         3000000   
998   500/1000               2000              1356.9200         5000000   
999    250/500               1000               766.1900               0   

     insured_zip insured_sex insured_education_level insured_occupation  \
995       431289      FEMALE                 Masters       craft-repair   
996       608177      FEMALE                     PhD     prof-specialty   
997       442797      FEMALE                 Masters       armed-forces   
998       441714        MALE               Associate  handlers-cleaners   
999       612260      FEMALE               Associate              sales   

    insured_hobbies insured_relationship  capital-gains  capital-loss  \
995       paintball            unmarried              0             0   
996        sleeping                 wife          70900             0   
997  bungie-jumping       other-relative          35100             0   
998    base-jumping                 wife              0             0   
999        kayaking              husband              0             0   

    incident_date             incident_type   collision_type  \
995    2015-02-22  Single Vehicle Collision  Front Collision   
996    2015-01-24  Single Vehicle Collision   Rear Collision   
997    2015-01-23   Multi-vehicle Collision   Side Collision   
998    2015-02-26  Single Vehicle Collision   Rear Collision   
999    2015-02-26                Parked Car                ?   

    incident_severity authorities_contacted incident_state incident_city  \
995      Minor Damage                  Fire             NC    Northbrook   
996      Major Damage                  Fire             SC     Northbend   
997      Minor Damage                Police             NC     Arlington   
998      Major Damage                 Other             NY     Arlington   
999      Minor Damage                Police             WV      Columbus   

       incident_location  incident_hour_of_the_day  \
995   6045 Andromedia St                        20   
996     3092 Texas Drive                        23   
997          7629 5th St                         4   
998        6128 Elm Lane                         2   
999  1416 Cherokee Ridge                         6   

     number_of_vehicles_involved property_damage  bodily_injuries  witnesses  \
995                            1             YES                0          1   
996                            1             YES                2          3   
997                            3               ?                2          3   
998                            1               ?                0          1   
999                            1               ?                0          3   

    police_report_available  total_claim_amount  injury_claim  property_claim  \
995                       ?               87200         17440            8720   
996                       ?              108480         18080           18080   
997                     YES               67500          7500            7500   
998                     YES               46980          5220            5220   
999                       ?                5060           460             920   

     vehicle_claim   auto_make auto_model  auto_year fraud_r

<span style="color:Chartreuse;font-size:120%;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> We can see that the few data points updated with ? rather than null. We must invetigate further.</span>

In [90]:
df.describe(include = "number").T

count         mean          std           min  \
months_as_customer          1000.0000     203.9540     115.1132        0.0000   
age                         1000.0000      38.9480       9.1403       19.0000   
policy_number               1000.0000  546238.6480  257063.0053   100804.0000   
policy_deductable           1000.0000    1136.0000     611.8647      500.0000   
policy_annual_premium       1000.0000    1256.4061     244.1674      433.3300   
umbrella_limit              1000.0000 1101000.0000 2297406.5981 -1000000.0000   
insured_zip                 1000.0000  501214.4880   71701.6109   430104.0000   
capital-gains               1000.0000   25126.1000   27872.1877        0.0000   
capital-loss                1000.0000  -26793.7000   28104.0967  -111100.0000   
incident_hour_of_the_day    1000.0000      11.6440       6.9514        0.0000   
number_of_vehicles_involved 1000.0000       1.8390       1.0189        1.0000   
bodily_injuries             1000.0000       0.9920       0.8201        0.0000   
witnesses                   1000.0000       1.4870       1.1113        0.0000   
total_claim_amount          1000.0000   52761.9400   26401.5332      100.0000   
injury_claim                1000.0000    7433.4200    4880.9519        0.0000   
property_claim              1000.0000    7399.5700    4824.7262        0.0000   
vehicle_claim               1000.0000   37928.9500   18886.2529       70.0000   
auto_year                   1000.0000    2005.1030       6.0159     1995.0000   

                                    25%         50%         75%           max  
months_as_customer             115.7500    199.5000    276.2500      479.0000  
age                             32.0000     38.0000     44.0000       64.0000  
policy_number               335980.2500 533135.0000 759099.7500   999435.0000  
policy_deductable              500.0000   1000.0000   2000.0000     2000.0000  
policy_annual_premium         1089.6075   1257.2000   1415.6950     2047.5900  
umbrella_limit                   0.0000      0.0000      0.0000 10000000.0000  
insured_zip                 448404.5000 466445.5000 603251.0000   620962.0000  
capital-gains                    0.0000      0.0000  51025.0000   100500.0000  
capital-loss                -51500.0000 -23250.0000      0.0000        0.0000  
incident_hour_of_the_day         6.0000     12.0000     17.0000       23.0000  
number_of_vehicles_involved      1.0000      1.0000      3.0000        4.0000  
bodily_injuries                  0.0000      1.0000      2.0000        2.0000  
witnesses                        1.0000      1.0000      2.0000        3.0000  
total_claim_amount           41812.5000  58055.0000  70592.5000   114920.0000  
injury_claim                  4295.0000   6775.0000  11305.0000    21450.0000  
property_claim                4445.0000   6750.0000  10885.0000    23670.0000  
vehicle_claim                30292.5000  42100.0000  50822.5000    79560.0000  
auto_year                     2000.0000   2005.0000   2010.0000     2015.0000

<span style="color:Chartreuse;font-size:120%;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> We can see the statistical characteristics of numerical features such as count, central tendency, standardeviation, minimum, maximum, and percentile value of each numerical features (columns). It will be helpful for us to understand the data distribution, structures, etc.</span>

In [91]:
df.describe(exclude = "number").T

count unique                      top freq  \
policy_bind_date         1000    NaN                      NaN  NaN   
policy_state             1000      3                       OH  352   
policy_csl               1000      3                  250/500  351   
insured_sex              1000      2                   FEMALE  537   
insured_education_level  1000      7                       JD  161   
insured_occupation       1000     14        machine-op-inspct   93   
insured_hobbies          1000     20                  reading   64   
insured_relationship     1000      6                own-child  183   
incident_date            1000    NaN                      NaN  NaN   
incident_type            1000      4  Multi-vehicle Collision  419   
collision_type           1000      4           Rear Collision  292   
incident_severity        1000      4             Minor Damage  354   
authorities_contacted     909      4                   Police  292   
incident_state           1000      7                       NY  262   
incident_city            1000      7              Springfield  157   
incident_location        1000   1000      1416 Cherokee Ridge    1   
property_damage          1000      3                        ?  360   
police_report_available  1000      3                        ?  343   
auto_make                1000     14                     Saab   80   
auto_model               1000     39                      RAM   43   
fraud_reported           1000      2                        N  753   

                                                  mean                  min  \
policy_bind_date         2002-02-08 04:40:47.999999872  1990-01-08 00:00:00   
policy_state                                       NaN                  NaN   
policy_csl                                         NaN                  NaN   
insured_sex                                        NaN                  NaN   
insured_education_level                            NaN                  NaN   
insured_occupation                                 NaN                  NaN   
insured_hobbies                                    NaN                  NaN   
insured_relationship                               NaN                  NaN   
incident_date                      2015-01-30 08:02:24  2015-01-01 00:00:00   
incident_type                                      NaN                  NaN   
collision_type                                     NaN                  NaN   
incident_severity                                  NaN                  NaN   
authorities_contacted                              NaN                  NaN   
incident_state                                     NaN                  NaN   
incident_city                                      NaN                  NaN   
incident_location                                  NaN                  NaN   
property_damage                                    NaN                  NaN   
police_report_available                            NaN                  NaN   
auto_make                                          NaN                  NaN   
auto_model                                         NaN                  NaN   
fraud_reported                                     NaN                  NaN   

                                         25%                  50%  \
policy_bind_date         1995-09-19 00:00:00  2002-04-01 12:00:00   
policy_state                             NaN                  NaN   
policy_csl                               NaN                  NaN   
insured_sex                              NaN                  NaN   
insured_education_level                  NaN                  NaN   
insured_occupation                       NaN                  NaN   
insured_hobbies                          NaN                  NaN   
insured_relationship                     NaN                  NaN   
incident_date            2015-01-15 00:00:00  2015-01-31 00:00:00   
incident_type                            NaN          

<span style="color:Chartreuse;font-size:120%;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> We can see the statistical characteristics of categorical features such as count, central tendency,unique values, frequency of each categorical features (columns). It will be helpful for us to understand the data distribution, stuctures, etc. </span>

### <span style="color:Khaki;">Feature Details:

In [92]:
df['police_report_available'].value_counts()

police_report_available
?      343
NO     343
YES    314
Name: count, dtype: int64

In [93]:
df.sample(3)

months_as_customer  age  policy_number policy_bind_date policy_state  \
739                 194   34         884365       1994-05-17           IN   
120                 255   41         368050       2013-01-08           IL   
544                 254   40         349658       1994-06-07           IN   

    policy_csl  policy_deductable  policy_annual_premium  umbrella_limit  \
739    100/300               1000              1423.3400               0   
120   500/1000               2000              1167.3000         4000000   
544    100/300                500              1927.8700               0   

     insured_zip insured_sex insured_education_level insured_occupation  \
739       454399        MALE               Associate              sales   
120       618123        MALE             High School    priv-house-serv   
544       609317        MALE                      MD     prof-specialty   

    insured_hobbies insured_relationship  capital-gains  capital-loss  \
739         camping        not-in-family          55300        -37900   
120     board-games       other-relative              0             0   
544        yachting              husband              0             0   

    incident_date             incident_type   collision_type  \
739    2015-01-21             Vehicle Theft                ?   
120    2015-02-22  Single Vehicle Collision   Side Collision   
544    2015-01-21  Single Vehicle Collision  Front Collision   

    incident_severity authorities_contacted incident_state incident_city  \
739      Minor Damage                   NaN             WV     Riverwood   
120      Minor Damage                  Fire             SC     Hillsdale   
544      Minor Damage                  Fire             VA     Arlington   

    incident_location  incident_hour_of_the_day  number_of_vehicles_involved  \
739       8453 Elm St                         0                            1   
120     8381 Solo Hwy                        22                            1   
544     6260 5th Lane                        10                            1   

    property_damage  bodily_injuries  witnesses police_report_available  \
739             YES                0          3                      NO   
120              NO                2          0                       ?   
544             YES                0          1                       ?   

     total_claim_amount  injury_claim  property_claim  vehicle_claim  \
739                9100          1400            1400           6300   
120               69300         13860           13860          41580   
544               43280             0            5410          37870   

      auto_make auto_model  auto_year fraud_reported  
739   Chevrolet     Malibu       2003              N  
120  Volkswagen     Passat       2000              N  
544       Honda      Civic       1996              Y

<span style="color:Tomato;font-size: 100%"> <b> 1. Month as  Customer - </b></span> Customer tenure ranges from new customers (0 months) to long-term ones (479 months), with a mean of about 204 months. Helps identify long-term vs new customers.

<span style="color:Tomato;font-size: 100%"> <b> 2. Age - </b></span> Important demographic factor and a crucial factor in premium calculation.

<span style="color:Tomato;font-size: 100%"> <b> 3. Policy Number - </b></span> Redundant features. Generally not used in modeling.

<span style="color:Tomato;font-size: 100%"> <b> 5. Policy Bind Date - </b></span> Policy start date. Can be used to calculate policy age.

<span style="color:Tomato;font-size: 100%"> <b> 6. Policy State - </b></span> Location. Location-based risk factors.

<span style="color:Tomato;font-size: 100%"> <b> 7. Policy CSL - </b></span>  Combined Single Limit. Indicates coverage level.

<span style="color:Tomato;font-size: 100%"> <b> 8. Policy Deductable - </b></span> Portion of a claim that policy holder responsible to pay. Redundant features. 

<span style="color:Tomato;font-size: 100%"> <b> 9. Policy Annual Premium - </b></span> Preminum amount indicates customer value. One of target feature.

<span style="color:Tomato;font-size: 100%"> <b> 10. Umbrella Limit - </b></span> Max coverage limit.

<span style="color:Tomato;font-size: 100%"> <b> 10. Insured Zip - </b></span> Zip/Postal codes can be used to determine the level of risk associated with an insured individual and also help in developing targeted insurance products.

<span style="color:Tomato;font-size: 100%"> <b> 11. Insured Sex - </b></span> Gender can help in developing targeted insurance products.

<span style="color:Tomato;font-size: 100%"> <b> 12. Insured Education Level - </b></span> Demographic information and proxy for risk and income.

<span style="color:Tomato;font-size: 100%"> <b> 13. Insured Occupation - </b></span> Lifestyle indicator Might reveal unexpected correlations with fraud.

<span style="color:Tomato;font-size: 100%"> <b> 14. Insured Hobbies - </b></span> Lifestyle and risk indicator. Might reveal unexpected correlations with fraud

<span style="color:Tomato;font-size: 100%"> <b> 15. Insured Relationship - </b></span> Family Status. May influence premium price, and could be a indicators of potential fraud.

<span style="color:Tomato;font-size: 100%"> <b> 16. Capital Gains and Loss - </b></span> Not sure about this feature. I will try interpret as we go on with EDA. If it is not helpful we can drop these features.

<span style="color:Tomato;font-size: 100%"> <b> 17. Incident Date - </b></span> Accident date, can Calculate policy age from policy bind date and incident date.

<span style="color:Tomato;font-size: 100%"> <b> 18. Incident Type and Collision Type - </b></span> Different types may have varying fraud rates. Past incidents may affect future premiums.

<span style="color:Tomato;font-size: 100%"> <b> 19. Incident Severity </b></span> Severe incidents might be less likely to be fraudulent.

<span style="color:Tomato;font-size: 100%"> <b> 20. Authorities Contacted - </b></span> This information can help in predicting claim amounts or categorizing claims based on severity. Also could be indicative of fraudulent claims.

<span style="color:Tomato;font-size: 100%"> <b> 21. Incident State and City - </b></span> Location-based risk factors.

<span style="color:Tomato;font-size: 100%"> <b> 22. Incident Hour - </b></span> Analyzing the relationship between incident hour and other variables (e.g., incident type, claim amount) can reveal patterns indicative of fraud.

<span style="color:Tomato;font-size: 100%"> <b> 23. Number Of Vehicle Involved - </b></span> This information can help in risk assessment and premium calculation.

<span style="color:Tomato;font-size: 100%"> <b> 24. Property Damage - </b></span> Property damage can be a factor in assessing risk for future insurance policies. Inconsistent information between property damage and other claim details might indicate potential fraud.

<span style="color:Tomato;font-size: 100%"> <b> 25. Bodily Injuries - </b></span> Unusual patterns in bodily injuries might indicate potential fraud.

<span style="color:Tomato;font-size: 100%"> <b> 26. Witnesses - </b></span> A higher number of witnesses could indicate a lower probability of fraud.

<span style="color:Tomato;font-size: 100%"> <b> 28. Police Report Available - </b></span> This feature can be a valuable indicator of potential fraud.

<span style="color:Tomato;font-size: 100%"> <b> 29. Claim Amount - </b></span> Can help identify anomalies.

<span style="color:Tomato;font-size: 100%"> <b> 30. Vehicle Details - </b></span> Premiums are often based on the vehicle's value, which is influenced by its make, model, and year. 

<span style="color:Tomato;font-size: 100%"> <b> 31. Fraud Reported - </b></span> One of our target variable.

### <span style="color:Khaki;"> Checking data Type:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the output below, we can see the data type of each feature. We can reduce memory usage by maintaining or casting applicable data types to the features. This process will smooth our data wrangling efforts.

In [94]:
df.dtypes

months_as_customer                      int64
age                                     int64
policy_number                           int64
policy_bind_date               datetime64[ns]
policy_state                           object
policy_csl                             object
policy_deductable                       int64
policy_annual_premium                 float64
umbrella_limit                          int64
insured_zip                             int64
insured_sex                            object
insured_education_level                object
insured_occupation                     object
insured_hobbies                        object
insured_relationship                   object
capital-gains                           int64
capital-loss                            int64
incident_date                  datetime64[ns]
incident_type                          object
collision_type                         object
incident_severity                      object
authorities_contacted             

In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 39 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   months_as_customer           1000 non-null   int64         
 1   age                          1000 non-null   int64         
 2   policy_number                1000 non-null   int64         
 3   policy_bind_date             1000 non-null   datetime64[ns]
 4   policy_state                 1000 non-null   object        
 5   policy_csl                   1000 non-null   object        
 6   policy_deductable            1000 non-null   int64         
 7   policy_annual_premium        1000 non-null   float64       
 8   umbrella_limit               1000 non-null   int64         
 9   insured_zip                  1000 non-null   int64         
 10  insured_sex                  1000 non-null   object        
 11  insured_education_level      1000 non-null  

### <span style="color:Khaki;">Replacing ? to Null:

In [96]:
df=df.replace('?',np.nan)

### <span style="color:Khaki;">Checking Null Values:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> This code checks and return the empty cells in our dataset. It is crusial to handle this null values to feef our data to Machine Learning algorithm.

In [97]:
print(df.isnull().sum().sum(),' -- ',df.isna().sum().sum())
df.isnull().sum()

972  --  972


months_as_customer               0
age                              0
policy_number                    0
policy_bind_date                 0
policy_state                     0
policy_csl                       0
policy_deductable                0
policy_annual_premium            0
umbrella_limit                   0
insured_zip                      0
insured_sex                      0
insured_education_level          0
insured_occupation               0
insured_hobbies                  0
insured_relationship             0
capital-gains                    0
capital-loss                     0
incident_date                    0
incident_type                    0
collision_type                 178
incident_severity                0
authorities_contacted           91
incident_state                   0
incident_city                    0
incident_location                0
incident_hour_of_the_day         0
number_of_vehicles_involved      0
property_damage                360
bodily_injuries     

### <span style="color:Khaki;">Checking unique values of the features:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> This code provides the unique datapoints that we have in the respective features and also checks for null values.

In [98]:
for col in df.columns:
    print(f"{col} - {len(df[col].unique())} - {df[col].nunique()} \n  Diff is {df[col].nunique() - len(df[col].unique())}") 

months_as_customer - 391 - 391 
  Diff is 0
age - 46 - 46 
  Diff is 0
policy_number - 1000 - 1000 
  Diff is 0
policy_bind_date - 951 - 951 
  Diff is 0
policy_state - 3 - 3 
  Diff is 0
policy_csl - 3 - 3 
  Diff is 0
policy_deductable - 3 - 3 
  Diff is 0
policy_annual_premium - 991 - 991 
  Diff is 0
umbrella_limit - 11 - 11 
  Diff is 0
insured_zip - 995 - 995 
  Diff is 0
insured_sex - 2 - 2 
  Diff is 0
insured_education_level - 7 - 7 
  Diff is 0
insured_occupation - 14 - 14 
  Diff is 0
insured_hobbies - 20 - 20 
  Diff is 0
insured_relationship - 6 - 6 
  Diff is 0
capital-gains - 338 - 338 
  Diff is 0
capital-loss - 354 - 354 
  Diff is 0
incident_date - 60 - 60 
  Diff is 0
incident_type - 4 - 4 
  Diff is 0
collision_type - 4 - 3 
  Diff is -1
incident_severity - 4 - 4 
  Diff is 0
authorities_contacted - 5 - 4 
  Diff is -1
incident_state - 7 - 7 
  Diff is 0
incident_city - 7 - 7 
  Diff is 0
incident_location - 1000 - 1000 
  Diff is 0
incident_hour_of_the_day - 24 - 2

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the output above, we can see that we have 12500 customer details in the overall 100000 data points.<br>
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Customer name count 10128 confirms that the customers may have multiple accounts.<br>
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Credit_Utilization_Ratio and ID are unique value. It may not be useful.

### <span style="color:Khaki;">Checking values lesser than or equal to 0:

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Sometimes the data may contain negative values or just updated with 0 due to data entry error. This code will show such datapoints.

In [99]:
for col in df.select_dtypes(include=['number']).columns:
    print(f"{col} --  {(df[col] <= 0).sum()}")

months_as_customer --  1
age --  0
policy_number --  0
policy_deductable --  0
policy_annual_premium --  0
umbrella_limit --  799
insured_zip --  0
capital-gains --  508
capital-loss --  1000
incident_hour_of_the_day --  52
number_of_vehicles_involved --  0
bodily_injuries --  340
witnesses --  249
total_claim_amount --  0
injury_claim --  25
property_claim --  19
vehicle_claim --  0
auto_year --  0


<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From above output, few of the features may have negative values or zero. We must check and address any errors.

In [100]:
df.columns

Index(['months_as_customer', 'age', 'policy_number', 'policy_bind_date',
       'policy_state', 'policy_csl', 'policy_deductable',
       'policy_annual_premium', 'umbrella_limit', 'insured_zip', 'insured_sex',
       'insured_education_level', 'insured_occupation', 'insured_hobbies',
       'insured_relationship', 'capital-gains', 'capital-loss',
       'incident_date', 'incident_type', 'collision_type', 'incident_severity',
       'authorities_contacted', 'incident_state', 'incident_city',
       'incident_location', 'incident_hour_of_the_day',
       'number_of_vehicles_involved', 'property_damage', 'bodily_injuries',
       'witnesses', 'police_report_available', 'total_claim_amount',
       'injury_claim', 'property_claim', 'vehicle_claim', 'auto_make',
       'auto_model', 'auto_year', 'fraud_reported'],
      dtype='object')

## <span style="color:Aqua;">Data Cleaning:

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> In this section we are gonna clean the data by imputing missing values etc.



### <span style="color:Khaki;">Imputing Null Values in collision_type:

In [101]:
print(df.collision_type.value_counts().values.sum())
df.collision_type.value_counts()

822


collision_type
Rear Collision     292
Side Collision     276
Front Collision    254
Name: count, dtype: int64

In [102]:
df['collision_type'] = df['collision_type'].fillna('Others')

In [103]:
print(df.collision_type.value_counts().values.sum())
df.collision_type.value_counts()

1000


collision_type
Rear Collision     292
Side Collision     276
Front Collision    254
Others             178
Name: count, dtype: int64

### <span style="color:Khaki;">Removing unecessary columns and updating column Names:

In [ ]:
df.columns # to print all columns names in our data

Index(['ID', 'Customer_ID', 'Month', 'Name', 'Age', 'SSN', 'Occupation',
       'Annual_Income', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts',
       'Num_Credit_Card', 'Interest_Rate', 'Num_of_Loan', 'Type_of_Loan',
       'Delay_from_due_date', 'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Mix', 'Outstanding_Debt',
       'Credit_Utilization_Ratio', 'Credit_History_Age',
       'Payment_of_Min_Amount', 'Total_EMI_per_month',
       'Amount_invested_monthly', 'Payment_Behaviour', 'Monthly_Balance',
       'Credit_Score'],
      dtype='object')

In [ ]:
df = df.rename(columns=str.lower) # changing column names to lower case for easy access
df.columns # printing column names to check 

Index(['id', 'customer_id', 'month', 'name', 'age', 'ssn', 'occupation',
       'annual_income', 'monthly_inhand_salary', 'num_bank_accounts',
       'num_credit_card', 'interest_rate', 'num_of_loan', 'type_of_loan',
       'delay_from_due_date', 'num_of_delayed_payment', 'changed_credit_limit',
       'num_credit_inquiries', 'credit_mix', 'outstanding_debt',
       'credit_utilization_ratio', 'credit_history_age',
       'payment_of_min_amount', 'total_emi_per_month',
       'amount_invested_monthly', 'payment_behaviour', 'monthly_balance',
       'credit_score'],
      dtype='object')

### <span style="color:Khaki;">Updating Data types:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Casting appropriate data types reduces dataset memeory and allow us to perform data cleaning operation smoothly.

In [ ]:
df.describe() # checking statistical information to cast appropriate datatype.

id  customer_id       month         age            ssn  \
count 100000.0000  100000.0000 100000.0000 100000.0000    100000.0000   
mean   80631.5000   25982.6666      4.5000     33.3163 500461680.2624   
std    43301.4866   14340.5431      2.2913     10.7648 290826734.3862   
min     5634.0000    1006.0000      1.0000     14.0000     81349.0000   
25%    43132.7500   13664.5000      2.7500     24.0000 245168577.2500   
50%    80631.5000   25777.0000      4.5000     33.0000 500688611.5000   
75%   118130.2500   38385.0000      6.2500     42.0000 756002666.2500   
max   155629.0000   50999.0000      8.0000     56.0000 999993421.0000   

       annual_income  monthly_inhand_salary  num_bank_accounts  \
count    100000.0000            100000.0000        100000.0000   
mean      50505.1234              4197.2708             5.3688   
std       38299.4221              3186.4325             2.5933   
min        7005.9300               303.6454             0.0000   
25%       19342.9725              1626.5942             3.0000   
50%       36999.7050              3095.9050             5.0000   
75%       71683.4700              5957.7150             7.0000   
max      179987.2800             15204.6333            11.0000   

       num_credit_card  interest_rate  num_of_loan  delay_from_due_date  \
count      100000.0000    100000.0000  100000.0000          100000.0000   
mean            5.5336        14.5321       3.5329              21.0814   
std             2.0671         8.7413       2.4464              14.8046   
min             0.0000         1.0000       0.0000               0.0000   
25%             4.0000         7.0000       2.0000              10.0000   
50%             5.0000        13.0000       3.0000              18.0000   
75%             7.0000        20.0000       5.0000              28.0000   
max            11.0000        34.0000       9.0000              62.0000   

       num_of_delayed_payment  changed_credit_limit  num_credit_inquiries  \
count             100000.0000           100000.0000           100000.0000   
mean                  13.3131               10.4703                5.7983   
std                    6.2372                6.6095                3.8678   
min                    0.0000                0.5000                0.0000   
25%                    9.0000                5.3800                3.0000   
50%                   14.0000                9.4000                5.0000   
75%                   18.0000               14.8500                8.0000   
max                   25.0000               29.9800               17.0000   

       outstanding_debt  credit_utilization_ratio  credit_history_age  \
count       100000.0000               100000.0000         100000.0000   
mean          1426.2204                   32.2852            221.2205   
std           1155.1290                    5.1169             99.6807   
min              0.2300                   20.0000              1.0000   
25%            566.0725                   28.0526            144.0000   
50%           1166.1550                   32.3058            219.0000   
75%           1945.9625                   36.4967            302.0000   
max           4998.0700                   50.0000            404.0000   

       total_emi_per_month  amount_invested_monthly  monthly_balance  
count          100000.0000              100000.0000      100000.0000  
mean              107.6992                  55.1013         392.6976  
std               132.2671                  39.0069         201.6527  
min                 0.0000                   0.0000           0.0078  
25%                29.2689                  27.9591         267.6160  
50%                66.4623                  45.1565         333.8654  
75%               147.3926                  71.2958         463.2157  
max              1779.1033                 434.1911        1183.9307

#### <span style="color:Khaki;">Updating data type for numeric features.

In [ ]:
dtype_dict = {
    'month': 'int8',
    'age': 'int8',
    'num_bank_accounts': 'int8',
    'num_credit_card': 'int8',
    'interest_rate': 'int8',
    'num_of_loan': 'int8',
    'delay_from_due_date': 'int8',
    'num_of_delayed_payment': 'int8',
    'changed_credit_limit': 'int8',
    'credit_utilization_ratio': 'int8',
    'credit_history_age': 'int32',
    'annual_income': 'float64',
    'monthly_inhand_salary': 'float64',
    'total_emi_per_month': 'float64',
    'amount_invested_monthly': 'float64',
    'monthly_balance': 'float64',
    'customer_id': 'int64'
}

df = df.astype(dtype_dict)

#### <span style="color:Khaki;">Updating data type for catagorical features.

In [ ]:
for col in df.select_dtypes(exclude='number').columns:
    df[col] = df[col].astype('category')

In [ ]:
df.info() # checking datatypes and memory usage after changing datatypes.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype   
---  ------                    --------------   -----   
 0   id                        100000 non-null  int64   
 1   customer_id               100000 non-null  int64   
 2   month                     100000 non-null  int8    
 3   name                      100000 non-null  category
 4   age                       100000 non-null  int8    
 5   ssn                       100000 non-null  float64 
 6   occupation                100000 non-null  category
 7   annual_income             100000 non-null  float64 
 8   monthly_inhand_salary     100000 non-null  float64 
 9   num_bank_accounts         100000 non-null  int8    
 10  num_credit_card           100000 non-null  int8    
 11  interest_rate             100000 non-null  int8    
 12  num_of_loan               100000 non-null  int8    
 13  type_of_loan              1000

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> After casting data type we could see that the memory usage is reduced to 8.8 MB.

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the above output we see

### <p style="color: Khaki;">Cleaning Case and Text inconsistency and checking unique values:</p>

In [ ]:
for col in df.select_dtypes(include='category'):
    if 'block' not in col:
        df[col] = df[col].str.strip().str.title()
        df[col].describe()
        print(df[col].describe(),'\n\n',df[col].value_counts().reset_index().sort_values(by=col,ascending=True),'\n','-----'*10)

count       100000
unique       10128
top       Jessicad
freq            48
Name: name, dtype: object 

                        name  count
8613              A Coopert      8
2526             A Maxwella      8
8192     A. Ananthalakshmih      8
10114    A. Ananthalakshmiq      8
4384     A. Ananthalakshmix      8
...                     ...    ...
3751             Zieminskis      8
7760             Zieminskiv      8
5525                Zinetsn      8
4248   Zoran Radosavljevicl      8
5938                 Zorans      8

[10128 rows x 2 columns] 
 --------------------------------------------------
count     100000
unique        15
top       Lawyer
freq        7096
Name: occupation, dtype: object 

        occupation  count
5      Accountant   6744
2       Architect   6824
6       Developer   6720
10         Doctor   6568
1        Engineer   6864
9    Entrepreneur   6648
11     Journalist   6536
0          Lawyer   7096
12        Manager   6432
3        Mechanic   6776
7   Media_Manager 

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> <span style="color: white;">From the above output, type_of_loan feature has array of values (clustered value). It will not be useful to our model as it recognizes this as unique text. So we are going to split each loan into each column to provide weightage based on the loan count. Finally, it creates columns with the all loan type and provide value to rows as 1 if someone has the loan or it will be 0 (One-Hot Encoding).

In [ ]:
# Splitting type of loans based on the comma

df['type_of_loan'] = df['type_of_loan'].apply(lambda x: x if isinstance(x, list) else x.split(','))

# Removing 'And' and white spaces 
df['type_of_loan'] = df['type_of_loan'].apply(lambda x: list(set([loan.replace('And ', '').strip()  for loan in x])))


In [ ]:
# One-Hot Encoding
loan_types = set(loan for loans in df['type_of_loan'] for loan in loans)
for loan_type in loan_types:
    df[loan_type] = df['type_of_loan'].apply(lambda x: 1 if loan_type in x else 0)
    
# Dropping original column and other loan type
df.drop(['type_of_loan','Not Specified','No Data'], axis=1, inplace=True)

In [ ]:
df.sample(2)

id  customer_id  month                name  age            ssn  \
67395  106725        48171      4       Jessica Wohlb   46 329376326.0000   
94964  148078         6383      5  Andrea Shalal-Esak   21 136445764.0000   

         occupation  annual_income  monthly_inhand_salary  num_bank_accounts  \
67395        Writer     20867.6700              1489.9725                  1   
94964  Entrepreneur    175127.0400             14635.9200                  4   

       num_credit_card  interest_rate  num_of_loan  delay_from_due_date  \
67395                3             11            1                   13   
94964                6             12            4                   15   

       num_of_delayed_payment  changed_credit_limit  num_credit_inquiries  \
67395                      14                     1                4.0000   
94964                       8                     8                3.0000   

      credit_mix  outstanding_debt  credit_utilization_ratio  \
67395       Good          884.5200                        38   
94964       Good          518.5500                        25   

       credit_history_age payment_of_min_amount  total_emi_per_month  \
67395                 378                    Nm              16.8372   
94964                 362                    No             421.3378   

       amount_invested_monthly                payment_behaviour  \
67395                   0.0000   Low_Spent_Small_Value_Payments   
94964                 125.4198  Low_Spent_Medium_Value_Payments   

       monthly_balance credit_score  Auto Loan  Credit-Builder Loan  \
67395         334.7935         Good          0                    0   
94964          59.1154     Standard          0                    0   

       Debt Consolidation Loan  Payday Loan  Home Equity Loan  Personal Loan  \
67395                        1            0                 0              0   
94964                        1            0                 1              0   

       Mortgage Loan  Student Loan  
67395              0             0  
94964              1             0

<p style="color: Khaki;">

## <p Style="color: Aqua"> Exporting Data Frame

In [ ]:
df.to_feather('Cleaned_data') # Exporting data frame as feather data type. It is efficient and keep our data types.